# Floating-Point Representation

In this section, we will discuss a commonly-occurring issue in which Python seemingly misrepresents the value of certain decimal numbers.

## Floating-Point Values

A **floating-point** value is a numerical value that contains a decimal point. For example, each of the following values are floating-point values:

    3.14159, 4.5, 2.3333333, 19.0

## Rounding Issues in Floating Point Calculations

We have seen examples in previous sections where Python seems to make very small rounding errors when providing the result of calculations involving floating-point values. 

For example, consider the result obtained from dividing 7 by 3. 

In [1]:
print(7 / 3)

2.3333333333333335


The exact value of $7/3$ is $2.3333...$, which the threes in the fractional part continuing on forever. Computers are not typically capable of storing decimal values with to infinite precision, and thus Python (and other programming languages) are forced to round values such as this. However, you would expect that the last digit of the rounded result would be a 3 rather than a 5. It would seem that Python is making a rounding error. 

This issue is perhaps demonstrated more strikingly in the next example, in which we add `0.1` and `0.2`.

In [2]:
print(0.1 + 0.2)

0.30000000000000004


You would obvsiously expect the result of the calculation above to be `0.3`, but Python reports a result that is very, very slightly larger. This is particularly strange considering that you would likely not expect this calculation to involve any rounding at all!

## Binary Representation

These unexpected results stem from the fact that computers internally represent numbers in a base 2 (binary) number system rather than the base 10 (decimal) system that modern humans are familiar with. We will not delve into the details of exactly why binary representations sometimes yield unexpected results when working with floating point numbers, but suffice it to say that tiny rounding "errors" can occur when converting a number from base 10 (as it appears on the screen) to base 2 (as it is stored by the computer), and vice-versa. You can read more about this topic here: [Floating-Point Error in Python](https://www.geeksforgeeks.org/floating-point-error-in-python/).





## How to Handle Such Errors

Understanding that operations involving float-point values might occasionally result in tiny rounding errors, the question remains: What do we do about it? In most cases, the answer is simple: Nothing. If you are using the result of such a calculation in another calculation, the value discrepency between the actual value and the calculated value is so small that it is unlikely to be relevant. 

If you are displaying the value as a final result, and find the incorrectly rounded result to be displeasing, then you can simply rounded the displayed result to any precision that is less than what was used to store the number. To see an illustration of this, we will again perform the calculation `0.1 + 0.2`, but will this time round the result to 10 decimal places. 

In [4]:
print(round(0.1 + 0.2, 10))

0.3


The key point to take away from this discussion is that you should understand that there is always a chance that a floating-point operation might produce a result that is slightly different from (but very close to) the exact result. 